workflow/vm/vm.go

```diff
@@ -0,0 +1,164 @@
+		// check if workflow matches with trigger created from event.
+		if ok := w.Match(t); ok {
+			// create a new chain, register it and create task execution request for
+			// its first task.
+			c := newChain(w)
+			ok, execReq := c.Next(e)
+			// chain might have just one task, if so, don't bother to add it chain list.
+			if !c.IsDone() {
+				v.executionChains = append(v.executionChains, c)
+			}
+			// check against empty workflow Tasks.
+			if ok {
+				v.ExecuctionRequests <- execReq
+			}
+		}
+	}
+}
+
+// Add adds a workflow to VM.
+func (v *VM) Add(w workflow.Workflow) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	v.workflows = append(v.workflows, w)
+}
+
+// Remove removes workflow and all of its execution chains from VM.
+func (v *VM) Remove(hash string) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	for i, w := range v.workflows {
+		if w.Hash == hash {
+			v.workflows = append(v.workflows[:i], v.workflows[i+1:]...)
+			for ii := len(v.executionChains) - 1; i >= 0; i-- {
+				if v.executionChains[ii].w.Hash == hash {
+					v.executionChains = append(v.executionChains[:ii], v.executionChains[ii+1:]...)
+				}
+			}
+			return
+		}
+	}
+}
+
+// Shutdown immediately ends vm and closes ExecutionRequests channel.
+func (v *VM) Shutdown() {
+	v.m.Lock()
+	defer v.m.Unlock()
+	v.closing = true
+	close(v.ExecuctionRequests)
+}
+
+// Event represents a new event/result input to the workflow VM.
+type Event struct {
+	// InstanceHash that event is emitted from.
+	InstanceHash string
+
+	// ParentHash of event if there is.
+	ParentHash []byte
+
+	// Key of event.
+	Key string
+
+	// TaskKey of event when event is emitted as a response to an execution.
+	TaskKey string
+
+	// Data is the payload of event.
+	Data map[string]interface{}
+
+	// Secret is generated by the Workflow VM for each execution request in a chain to check which
+	// Event belongs to which Execution request's result.
+	Secret string
+}
+
+// createTrigger creates a new trigger schema from event which is used to match the
+// event with registered workflows in VM.
+func (e Event) createTrigger() workflow.Trigger {
+	return workflow.Trigger{
+		EventKey:     e.Key,
+		InstanceHash: e.InstanceHash,
+		Filter: workflow.Filter{
+			TaskKey: e.TaskKey,
+		},
+	}
+}
+
+// Execution is an execution request that should be performed by the caller.
+// Each received Event from the Process() may cause multiple Execution requests
+// to be emitted by the Workflow VM.
+type Execution struct {
+	// InstanceHash of instance that execution should be made on.
+	InstanceHash string
+
+	// ParentHash is the parent execution hash if execution has one.
+	ParentHash []byte
+
+	// TaskKey of execution.
+	TaskKey string
+
+	// Inputs of the execution.
+	Inputs map[string]interface{}
+
+	// Secret is generated by the Workflow VM for each execution request in a chain to check which
+	// Event belongs to which Execution request's result.
+	Secret string
```

Comment:
If secret is used to identify Execution why not to Use execution hash and keep map of this in vm, instead add o new variable. 

Also if we want to add new variable then it should be called something with ID or Hash, as secret is rather for passwords,passpharse etc...

Reply:
workflowvm does not and cannot now about execution ids. execution id is generated after execution is actually made. here, we're just creating the execution requests. this is the same logic why we've introduced execution tags since execution ids weren't just enough.

this Secret is not a new variable. it should be converted to an execution tag by the caller of workflowvm.